### PR TITLE
feat(INT-555): add batch step executor

### PR DESCRIPTION
**Pull request by koladilip**
## Description of the change

Adds batch step executor for natively support batching workflow without end users writing any code

### Development

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -409,6 +437,25 @@ Supports importing of external functions and data using bindings.
     
     - In the above example: **processECommerace** step is the workflow step and importing additional bindings. Both workflow’s (**commonBinding**) and step’s (**stepBinding**) bindings are available to the workflow step.
 
+### Batch Step
+This helps to batch the inputs by length or size. We can also define our own batching by implementing [**BatchExecutor**](./src/steps/types.ts#L114) interface.
+
```
Comment (utsabc) [inline README.md:442]: Based on filter the key also will be returned with the batched output we can mention this as well

File: src/common/utils/batch.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      items.push(currentChunk);
+      indices.push(currentIndices);
+    }
+
+    return { items, indices };
+  }
+
+  static validateBatchResults(input: any[], batchResults: BatchResult[]) {
+    if (!Array.isArray(batchResults)) {
+      throw new BatchError('batch step requires array output from batch executor');
```
Comment (utsabc) [inline src/common/utils/batch.ts:46]: What will be the structure of BatchStepOutput for this error?

File: src/steps/base/batch/default_batch_workflow_executor.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { ExecutionBindings } from '../../../workflow';
+import { BatchExecutor, BatchResult } from '../../../steps/types';
+
+export class DefaultBatchWorkflowExecutor implements BatchExecutor {
```
Comment (utsabc) [inline src/steps/base/batch/default_batch_workflow_executor.ts:4]: SHould we call this default considering its doing custom batching logic?

File: src/common/utils/batch.ts (hunk 2 above)
Comment (koladilip) [reply src/common/utils/batch.ts:46]: whole execution fails if this happens so no need to capture in BatchStepOutput

File: src/steps/base/batch/default_batch_workflow_executor.ts (hunk 3 above)
Comment (koladilip) [reply src/steps/base/batch/default_batch_workflow_executor.ts:4]: what do you mean by custom batching logic here

**Review by utsabc — APPROVED** [review]

**Pull request merged**